include/sisl/cache/simple_cache.hpp

```diff
@@ -42,14 +43,27 @@ class SimpleCache {
             m_key_extract_cb{std::move(extract_cb)},
             m_map{num_buckets, m_key_extract_cb, std::bind(&SimpleCache< K, V >::on_hash_operation, this, _1, _2, _3)},
             m_per_value_size{per_val_size} {
-        m_record_family_id = m_evictor->register_record_family(std::move(evict_cb));
+        m_record_family_id = m_evictor->register_record_family(std::move(evict_cb), [this](const CacheRecord& record) {
+            V const value = reinterpret_cast<SingleEntryHashNode< V >*>(const_cast< CacheRecord* >(&record))->m_value;
+            K key = m_key_extract_cb(value);
+            t_evicted_keys.insert(key);
+        });
     }
 
     ~SimpleCache() { m_evictor->unregister_record_family(m_record_family_id); }
 
     bool insert(const V& value) {
         K k = m_key_extract_cb(value);
-        return m_map.insert(k, value);
+        bool ret = m_map.insert(k, value);
+        // erase evicted keys
+        if (t_evicted_keys.size() > 0) {
+            for (const auto& key : t_evicted_keys) {
+                V out_val;
+                m_map.erase(key, out_val);
```

Comment:
here will be a case that if a thread evicts some keys , but `insert` is never called in this thread from then on,  then those keys will not be really evicted eventually.

so we need make sure the thread , which does the eviction, will call insert some day in the future

Reply:
I did not get the question. The keys are evicted as a part of the insert. When the insert call is done in the simple cache, the evicted keys list is empty

Reply:
`thread_local std::set< K > SimpleCache< K, V >::t_evicted_keys;`

this is a thread local variable. so if we remove a key in thread 1, then only this thread knows that this key is in the  t_evicted_keys and pending to be removed in the next `insert` call, right?

if a thread removes a key , and then exit without calling any `insert`, the this key will not be aware by other thread, and will never be removed from m_map.

not sure if I understand correctly , pls correct me if I am wrong